provider/workspace_preset_test.go

```diff
@@ -265,6 +265,249 @@ func TestWorkspacePreset(t *testing.T) {
+							instances = 3
+					  	}
+					}
+				}
+			}`,
+			ExpectError: regexp.MustCompile(`The argument "timezone" is required, but no definition was found.`),
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling field, but without schedule",
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				prebuilds {
+					instances = 1
+					autoscaling {
+						timezone = "UTC"
+					}
+				}
+			}`,
+			ExpectError: regexp.MustCompile(`At least 1 "schedule" blocks are required.`),
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling.schedule field, but without cron",
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				prebuilds {
+					instances = 1
+					autoscaling {
+						timezone = "UTC"
+						schedule {
+							instances = 3
+					  	}
+					}
+				}
+			}`,
+			ExpectError: regexp.MustCompile(`The argument "cron" is required, but no definition was found.`),
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling.schedule field, but without instances",
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				prebuilds {
+					instances = 1
+					autoscaling {
+						timezone = "UTC"
+						schedule {
+							cron = "* 8-18 * * 1-5"
+					  	}
+					}
+				}
+			}`,
+			ExpectError: regexp.MustCompile(`The argument "instances" is required, but no definition was found.`),
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling.schedule field, but with invalid type for instances",
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				prebuilds {
+					instances = 1
+					autoscaling {
+						timezone = "UTC"
+						schedule {
+							cron = "* 8-18 * * 1-5"
+							instances = "not_a_number"
+					  	}
+					}
+				}
+			}`,
+			ExpectError: regexp.MustCompile(`Inappropriate value for attribute "instances": a number is required`),
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling field with 1 schedule",
+			Config: `
+			data "coder_workspace_preset" "preset_1" {
+				name = "preset_1"
+				prebuilds {
+					instances = 1
+					autoscaling {
+						timezone = "UTC"
+					  	schedule {
+							cron = "* 8-18 * * 1-5"
+							instances = 3
+					  	}
+					}
+				}
+			}`,
+			ExpectError: nil,
+			Check: func(state *terraform.State) error {
+				require.Len(t, state.Modules, 1)
+				require.Len(t, state.Modules[0].Resources, 1)
+				resource := state.Modules[0].Resources["data.coder_workspace_preset.preset_1"]
+				require.NotNil(t, resource)
+				attrs := resource.Primary.Attributes
+				require.Equal(t, attrs["name"], "preset_1")
+				require.Equal(t, attrs["prebuilds.0.autoscaling.0.timezone"], "UTC")
+				require.Equal(t, attrs["prebuilds.0.autoscaling.0.schedule.0.cron"], "* 8-18 * * 1-5")
+				require.Equal(t, attrs["prebuilds.0.autoscaling.0.schedule.0.instances"], "3")
+				return nil
+			},
+		},
+		{
+			Name: "Prebuilds is set with an autoscaling field with 2 schedules",
```

Comment:
Where do we check that these schedules are not conflicting? I didn't find any checks in https://github.com/coder/coder/pull/18126 from a quick scan. I'll review that PR in more detail tomorrow.

Reply:
> If the current time matches any defined autoscaling schedule, the corresponding number of instances is used.

This makes me nervous because it introduces some non-determinism in the case of conflicting/overlapping schedules. If we don't evaluate cron schedules in the same order, we could flip back and forth between conflicting schedules then it could cause a lot of churn.

Reply:
Fixed. I added validation for overlapping schedules. It became a bit complex, considering the fact that DOM and months are allowed, and considering the fact that there are some tricky behavior when it comes to DOM, and DOW used together.